framework/src/meshmodifiers/BreakMeshByBlock.C

```diff
@@ -50,91 +50,96 @@ BreakMeshByBlock::modify()
+                       // loop
+              }
+
+            for (auto connected_elem_id : connected_elems)
             {
-              // add the current element to the neighbor elements list
-              neighbor_elems.insert(current_elem->id());
-
-              // fix current element node:
-              // cicle over the neighbor element nodes to find the node to fix
-              for (unsigned int local_elem_node_it = 0;
-                   local_elem_node_it < current_elem->n_nodes();
-                   ++local_elem_node_it)
-                if (current_elem->node_id(local_elem_node_it) ==
-                    current_node->id()) // if current node == node on element
-                {
-                  current_elem->set_node(local_elem_node_it) = new_node;
-                  break; // ones the proper node has been fixed in one element we can break the loop
-                }
-            } // end reference/neighbor actions
-          }   // end loop over connected elements
-        }     // end multiplicity loop
-
-        // to add the appropriate side to the new interface we need to loop over
-        // reference elements and find all the neighbors sharing a side with it
-        for (auto ref_elem_id : reference_elems)
-        {
-
-          Elem * ref_elem = mesh.elem_ptr(ref_elem_id);
+              Elem * connected_elem = mesh.elem_ptr(connected_elem_id);
+
+              // Assign the newly added node to other connected elements with the same block_id
+              if (connected_elem->subdomain_id() == current_elem->subdomain_id() &&
+                  connected_elem != current_elem)
+              {
+                for (unsigned int local_elem_node_it = 0;
+                     local_elem_node_it < connected_elem->n_nodes();
+                     ++local_elem_node_it)
+                  if (connected_elem->node_id(local_elem_node_it) ==
+                      current_node->id()) // if current node == node on element
+                  {
+                    connected_elem->set_node(local_elem_node_it) = new_node;
+                    break;
+                  }
+              }
+            }
+          }
+        }
 
-          // loop over possible neighbors
-          for (auto neigh_elem_id : neighbor_elems)
+        // create blocks pair and assign element side to new interface boundary map
+        for (auto elem_id : connected_elems)
+        {
+          for (auto connected_elem_id : connected_elems)
           {
-            Elem * neigh_elem = mesh.elem_ptr(neigh_elem_id);
-            if (ref_elem->has_neighbor(neigh_elem))
-            {
-              std::pair<subdomain_id_type, subdomain_id_type> blocks_pair =
-                  std::make_pair(ref_elem->subdomain_id(), neigh_elem->subdomain_id());
+            Elem * current_elem = mesh.elem_ptr(elem_id);
+            Elem * connected_elem = mesh.elem_ptr(connected_elem_id);
```

Comment:
You really should start asserting that you actually get these back as opposed to a nullptr or a remote elem. On Distributed mesh, you might not get a usable element. Since this object is currently restricted to replicated only, no action is required here. Just noting this for you.

Reply:
Thanks for this note :-)